File: f1tenth_lidar/src/hokuyo_scip_driver.py
#!/usr/bin/env python3
"""
Hokuyo UST-10LX SCIP 2.0 Direct Driver for ROS 2

Uses the SCIP 2.0 MD command for continuous streaming at the full 40 Hz
sensor rate, bypassing urg_node's synchronous request/reply limitation.

Protocol reference: Hokuyo SCIP 2.0 specification

Usage:
  ros2 run f1tenth_localization hokuyo_scip_driver.py
  ros2 run f1tenth_localization hokuyo_scip_driver.py --ros-args -p ip_address:=192.168.0.10
"""

import math
import socket
import struct
import threading
import time
from typing import Optional

import rclpy
from rclpy.node import Node
from rclpy.qos import QoSProfile, ReliabilityPolicy, DurabilityPolicy
from sensor_msgs.msg import LaserScan


def scip_decode(data: str) -> list[int]:
    """Decode SCIP 2.0 character-encoded data (3-character encoding)."""
    values = []
    # Remove line feeds within data block and checksum characters
    clean = ""
    for line in data.split("\n"):
        if len(line) > 0:
            # Last char of each line is checksum, strip it
            clean += line[:-1]

    for i in range(0, len(clean), 3):
        if i + 3 <= len(clean):
            # 3-character encoding: each char encodes 6 bits
            val = 0
            for j in range(3):
                val = (val << 6) + (ord(clean[i + j]) - 0x30)
            values.append(val)
    return values


class HokuyoScipDriver(Node):
    """Direct SCIP 2.0 driver for Hokuyo UST-10LX at full 40 Hz."""

    def __init__(self):
        super().__init__("hokuyo_scip_driver")

        # All parameters are loaded from config/hokuyo_ust10lx.yaml via the
        # launch file.  Defaults here are a safety-net mirror of the YAML.
        self.declare_parameter("ip_address", "192.168.0.10")
        self.declare_parameter("ip_port", 10940)
        self.declare_parameter("laser_frame_id", "ego_racecar/laser")
        self.declare_parameter("angle_min", -2.356194)  # -135 deg
        self.declare_parameter("angle_max", 2.356194)   # +135 deg
        self.declare_parameter("range_min", 0.1)
        self.declare_parameter("range_max", 10.0)
        self.declare_parameter("scan_topic", "/scan")
        self.declare_parameter("cluster", 4)   # 4 → 270 beams at 1° resolution
        self.declare_parameter("skip", 0)      # 0 → every scan (full 40 Hz)

        self.ip_address = self.get_parameter("ip_address").value
        self.ip_port = self.get_parameter("ip_port").value
        self.frame_id = self.get_parameter("laser_frame_id").value
        self.angle_min = self.get_parameter("angle_min").value
        self.angle_max = self.get_parameter("angle_max").value
        self.range_min = self.get_parameter("range_min").value
        self.range_max = self.get_parameter("range_max").value
        scan_topic = self.get_parameter("scan_topic").value
        self.cluster = self.get_parameter("cluster").value
        self.skip = self.get_parameter("skip").value

        # UST-10LX sensor specs
        self.total_steps = 1080  # Total measurement steps
        self.step_min = 0
        self.step_max = 1080
        self.angular_resolution = math.radians(0.25)  # 0.25 degrees per step
        self.scan_time = 0.100  # 25ms = 40 Hz

        # Publisher with SensorDataQoS (best-effort for real-time sensor data)
        sensor_qos = QoSProfile(
            reliability=ReliabilityPolicy.RELIABLE,
            durability=DurabilityPolicy.VOLATILE,
            depth=5,
        )
        self.scan_pub = self.create_publisher(LaserScan, scan_topic, sensor_qos)

        # Connection state
        self.sock: Optional[socket.socket] = None
        self.running = False
        self.recv_thread: Optional[threading.Thread] = None

        self.get_logger().info(f"Hokuyo SCIP 2.0 Direct Driver")
        self.get_logger().info(f"  Target: {self.ip_address}:{self.ip_port}")
        self.get_logger().info(f"  Frame: {self.frame_id}")
        self.get_logger().info(f"  Topic: {scan_topic}")
        self.get_logger().info(f"  Cluster: {self.cluster}, Skip: {self.skip}")

        # Connect and start streaming
        if self.connect():
            self.start_streaming()

    def connect(self) -> bool:
        """Connect to the sensor via TCP."""
        try:
            self.sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            self.sock.settimeout(5.0)
            self.sock.connect((self.ip_address, self.ip_port))
            self.get_logger().info("Connected to sensor")

            # Switch to SCIP 2.0 mode
            self._send_command("SCIP2.0\n")
            time.sleep(0.1)
            self._read_response()

            # Query sensor info
            self._send_command("VV\n")
            time.sleep(0.1)
            response = self._read_response()
            for line in response.split("\n"):
                if line.startswith("PROD:") or line.startswith("FIRM:") or line.startswith("SERI:"):
                    self.get_logger().info(f"  {line.strip()}")

            # Query sensor parameters
            self._send_command("PP\n")
            time.sleep(0.1)
            pp_response = self._read_response()
            self._parse_sensor_params(pp_response)

            return True
        except Exception as e:
            self.get_logger().error(f"Failed to connect: {e}")
            return False

    def _parse_sensor_params(self, response: str):
        """Parse PP response for sensor parameters."""
        for line in response.split("\n"):
            line = line.strip()
            if line.startswith("AMIN:"):
                self.step_min = int(line.split(":")[1].split(";")[0])
            elif line.startswith("AMAX:"):
                self.step_max = int(line.split(":")[1].split(";")[0])
            elif line.startswith("ARES:"):
                total_steps = int(line.split(":")[1].split(";")[0])
                if total_steps > 0:
                    self.angular_resolution = (2.0 * math.pi) / total_steps
            elif line.startswith("SCAN:"):
                scan_rpm = int(line.split(":")[1].split(";")[0])
                if scan_rpm > 0:
                    self.scan_time = 60.0 / scan_rpm

        num_steps = self.step_max - self.step_min
        self.get_logger().info(
            f"  Sensor: steps {self.step_min}-{self.step_max} "
            f"({num_steps} points), scan time {self.scan_time * 1000:.1f}ms"
        )

    def _send_command(self, cmd: str):
        """Send a SCIP command."""
        if self.sock:
            self.sock.sendall(cmd.encode("ascii"))

    def _read_response(self) -> str:
        """Read a complete SCIP response (terminated by double LF)."""
        buf = b""
        self.sock.settimeout(2.0)
        try:
            while True:
                chunk = self.sock.recv(4096)
                if not chunk:
                    break
                buf += chunk
                # SCIP responses end with \n\n
                if b"\n\n" in buf:
                    break
        except socket.timeout:
            pass
        return buf.decode("ascii", errors="replace")

    def start_streaming(self):
        """Start continuous distance measurement using MD command."""
        # MD command format: MDsssseeeecc0ii\n
        #   ssss = start step (4 digits)
        #   eeee = end step (4 digits)
        #   cc   = cluster count (2 digits)
        #   0    = scan interval (0 = every scan)
        #   ii   = number of scans (00 = infinite)
        start = self.step_min
        end = self.step_max
        cluster = self.cluster
        skip = self.skip

        cmd = f"MD{start:04d}{end:04d}{cluster:02d}{skip:01d}00\n"
        self.get_logger().info(f"Starting continuous stream: {cmd.strip()}")

        self._send_command(cmd)

        # Read the echo/status response
        time.sleep(0.1)
        status = self._read_response()
        status_lines = [l for l in status.split("\n") if l.strip()]
        if len(status_lines) >= 2:
            status_code = status_lines[1][:2] if len(status_lines[1]) >= 2 else "??"
            if status_code == "00" or status_code == "99":
                self.get_logger().info("Continuous streaming started (40 Hz)")
            else:
                self.get_logger().warn(f"MD status: {status_code} — {status}")

        # Start receive thread
        self.running = True
        self.sock.settimeout(1.0)
        self.recv_thread = threading.Thread(target=self._receive_loop, daemon=True)
        self.recv_thread.start()

    def _receive_loop(self):
        """Continuously receive and parse streaming scan data."""
        buf = b""
        while self.running and rclpy.ok():
            try:
                chunk = self.sock.recv(8192)
                if not chunk:
                    self.get_logger().error("Connection lost")
                    break
                buf += chunk

                # Process complete messages (separated by \n\n)
                while b"\n\n" in buf:
                    idx = buf.index(b"\n\n")
                    message = buf[:idx].decode("ascii", errors="replace")
                    buf = buf[idx + 2 :]
                    self._process_md_response(message)

            except socket.timeout:
                continue
            except Exception as e:
                if self.running:
                    self.get_logger().error(f"Receive error: {e}")
                break

    def _process_md_response(self, message: str):
        """Parse an MD streaming response and publish LaserScan."""
        lines = message.split("\n")
        if len(lines) < 3:
            return

        # First line is the echo of the MD command
        header = lines[0]
        if not header.startswith("MD"):
            return

        # Second line is status + timestamp
        status_line = lines[1]
        if len(status_line) < 2:
            return
        status = status_line[:2]
        if status != "99":
            # 99 = data follows, 00 = command accepted (initial response)
            return

        # Third line is timestamp (4-byte encoded)
        if len(lines) < 4:
            return
        timestamp_line = lines[2]

        # Remaining lines are distance data (3-char encoded, checksum at end of each line)
        data_str = "\n".join(lines[3:])
        if not data_str.strip():
            return

        try:
            distances_mm = scip_decode(data_str)
        except Exception as e:
            self.get_logger().debug(f"Decode error: {e}")
            return

        if len(distances_mm) == 0:
            return

        # Build LaserScan message
        scan = LaserScan()
        now = self.get_clock().now()
        scan.header.stamp = now.to_msg()
        scan.header.frame_id = self.frame_id

        num_points = len(distances_mm)
        effective_resolution = self.angular_resolution * self.cluster

        # Calculate angle range based on actual step range
        total_angle = effective_resolution * (num_points - 1)
        center_step = (self.step_max + self.step_min) / 2.0
        half_steps = (self.step_max - self.step_min) / 2.0
        
        scan.angle_min = -half_steps * self.angular_resolution
        scan.angle_max = half_steps * self.angular_resolution
        scan.angle_increment = effective_resolution
        scan.time_increment = self.scan_time / num_points
        scan.scan_time = self.scan_time
        scan.range_min = self.range_min
        scan.range_max = self.range_max

        # Convert mm to meters, apply range filtering
        scan.ranges = []
        for d in distances_mm:
            if d <= 20:  # Invalid reading (< 20mm typically means error)
                scan.ranges.append(float("inf"))
            else:
                r = d / 1000.0  # mm to meters
                if r < self.range_min or r > self.range_max:
                    scan.ranges.append(float("inf"))
                else:
                    scan.ranges.append(r)

        self.scan_pub.publish(scan)

    def stop_streaming(self):
        """Stop continuous measurement."""
        self.running = False
        if self.sock:
            try:
                # QT command stops streaming
                self._send_command("QT\n")
                time.sleep(0.1)
            except Exception:
                pass

    def destroy_node(self):
        """Clean shutdown."""
        self.stop_streaming()
        if self.sock:
            try:
                self.sock.close()
            except Exception:
                pass
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = HokuyoScipDriver()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.try_shutdown()


if __name__ == "__main__":
    main()

File: f1tenth_system/f1tenth_lidar/src/hokuyo_scip_driver.cpp
// Copyright (c) 2025 Pascal — MIT License
#include "f1tenth_lidar/hokuyo_scip_driver.hpp"

#include <arpa/inet.h>
#include <netinet/tcp.h>
#include <sys/socket.h>
#include <unistd.h>

#include <algorithm>
#include <chrono>
#include <cmath>
#include <cstring>
#include <sstream>
#include <stdexcept>

namespace f1tenth_lidar
{

// ═════════════════════════════════════════════════════════════════════════════
//  Construction / Destruction
// ═════════════════════════════════════════════════════════════════════════════

HokuyoScipDriver::HokuyoScipDriver(const rclcpp::NodeOptions & options)
: Node("hokuyo_scip_driver", options)
{
  // Declare parameters (defaults mirror config/hokuyo_ust10lx.yaml)
  declare_parameter("ip_address", std::string("192.168.10.10"));
  declare_parameter("ip_port", 10940);
  declare_parameter("laser_frame_id", std::string("ego_racecar/laser"));
  declare_parameter("angle_min", -2.356194);
  declare_parameter("angle_max",  2.356194);
  declare_parameter("range_min", 0.1);
  declare_parameter("range_max", 10.0);
  declare_parameter("scan_topic", std::string("/scan"));
  declare_parameter("cluster", 4);
  declare_parameter("skip", 0);

  ip_address_ = get_parameter("ip_address").as_string();
  ip_port_    = get_parameter("ip_port").as_int();
  frame_id_   = get_parameter("laser_frame_id").as_string();
  angle_min_  = get_parameter("angle_min").as_double();
  angle_max_  = get_parameter("angle_max").as_double();
  range_min_  = get_parameter("range_min").as_double();
  range_max_  = get_parameter("range_max").as_double();
  cluster_    = get_parameter("cluster").as_int();
  skip_       = get_parameter("skip").as_int();
  const auto scan_topic = get_parameter("scan_topic").as_string();

  // Publisher — RELIABLE / VOLATILE / depth 5 (matches Python driver)
  auto qos = rclcpp::QoS(5)
    .reliable()
    .durability_volatile();
  scan_pub_ = create_publisher<sensor_msgs::msg::LaserScan>(scan_topic, qos);

  RCLCPP_INFO(get_logger(), "Hokuyo SCIP 2.0 Direct Driver (C++)");
  RCLCPP_INFO(get_logger(), "  Target: %s:%d", ip_address_.c_str(), ip_port_);
  RCLCPP_INFO(get_logger(), "  Frame:  %s", frame_id_.c_str());
  RCLCPP_INFO(get_logger(), "  Topic:  %s", scan_topic.c_str());
  RCLCPP_INFO(get_logger(), "  Cluster: %d,  Skip: %d", cluster_, skip_);

  if (connect()) {
    start_streaming();
  }
}

HokuyoScipDriver::~HokuyoScipDriver()
{
  running_ = false;
  if (recv_thread_.joinable()) {
    recv_thread_.join();
  }
  if (sock_fd_ >= 0) {
    // Send QT to stop streaming before closing
    try { send_command("QT\n"); } catch (...) {}
    ::close(sock_fd_);
    sock_fd_ = -1;
  }
}

// ═════════════════════════════════════════════════════════════════════════════
//  TCP Connection
// ═════════════════════════════════════════════════════════════════════════════

bool HokuyoScipDriver::connect()
{
  sock_fd_ = ::socket(AF_INET, SOCK_STREAM, 0);
  if (sock_fd_ < 0) {
    RCLCPP_ERROR(get_logger(), "socket() failed: %s", std::strerror(errno));
    return false;
  }

  // Disable Nagle for snappier command exchange
  int flag = 1;
  ::setsockopt(sock_fd_, IPPROTO_TCP, TCP_NODELAY, &flag, sizeof(flag));

  // Connect with timeout
  struct sockaddr_in addr{};
  addr.sin_family = AF_INET;
  addr.sin_port   = htons(static_cast<uint16_t>(ip_port_));
  if (::inet_pton(AF_INET, ip_address_.c_str(), &addr.sin_addr) != 1) {
    RCLCPP_ERROR(get_logger(), "Invalid IP: %s", ip_address_.c_str());
    ::close(sock_fd_);
    sock_fd_ = -1;
    return false;
  }

  struct timeval tv{5, 0};  // 5 s connect timeout
  ::setsockopt(sock_fd_, SOL_SOCKET, SO_SNDTIMEO, &tv, sizeof(tv));
  ::setsockopt(sock_fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

  if (::connect(sock_fd_, reinterpret_cast<sockaddr *>(&addr), sizeof(addr)) != 0) {
    RCLCPP_ERROR(get_logger(), "connect() failed: %s", std::strerror(errno));
    ::close(sock_fd_);
    sock_fd_ = -1;
    return false;
  }
  RCLCPP_INFO(get_logger(), "Connected to sensor");

  // Switch to SCIP 2.0 mode
  send_command("SCIP2.0\n");
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  read_response();

  // Query sensor info (VV)
  send_command("VV\n");
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  const auto vv = read_response();
  std::istringstream vv_stream(vv);
  std::string line;
  while (std::getline(vv_stream, line)) {
    if (line.rfind("PROD:", 0) == 0 ||
        line.rfind("FIRM:", 0) == 0 ||
        line.rfind("SERI:", 0) == 0) {
      // Trim trailing whitespace
      while (!line.empty() && (line.back() == '\r' || line.back() == '\n' || line.back() == ' '))
        line.pop_back();
      RCLCPP_INFO(get_logger(), "  %s", line.c_str());
    }
  }

  // Query sensor parameters (PP)
  send_command("PP\n");
  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  const auto pp = read_response();
  parse_sensor_params(pp);

  return true;
}

void HokuyoScipDriver::send_command(const std::string & cmd)
{
  if (sock_fd_ < 0) return;
  ssize_t n = ::send(sock_fd_, cmd.data(), cmd.size(), MSG_NOSIGNAL);
  if (n < 0) {
    RCLCPP_WARN(get_logger(), "send() failed: %s", std::strerror(errno));
  }
}

std::string HokuyoScipDriver::read_response(double timeout_sec)
{
  std::string buf;
  buf.reserve(4096);

  struct timeval tv{};
  tv.tv_sec  = static_cast<long>(timeout_sec);
  tv.tv_usec = static_cast<long>((timeout_sec - tv.tv_sec) * 1e6);
  ::setsockopt(sock_fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

  char chunk[4096];
  while (true) {
    ssize_t n = ::recv(sock_fd_, chunk, sizeof(chunk), 0);
    if (n <= 0) break;
    buf.append(chunk, static_cast<size_t>(n));
    // SCIP responses end with "\n\n"
    if (buf.size() >= 2 && buf[buf.size() - 1] == '\n' && buf[buf.size() - 2] == '\n') {
      break;
    }
  }
  return buf;
}

void HokuyoScipDriver::parse_sensor_params(const std::string & response)
{
  std::istringstream ss(response);
  std::string line;
  while (std::getline(ss, line)) {
    // Remove trailing CR / whitespace
    while (!line.empty() && (line.back() == '\r' || line.back() == ' '))
      line.pop_back();

    auto value_of = [&](const std::string & prefix) -> int {
      if (line.rfind(prefix, 0) != 0) return -1;
      auto pos = line.find(':');
      if (pos == std::string::npos) return -1;
      auto semi = line.find(';', pos);
      std::string val_str = (semi != std::string::npos)
        ? line.substr(pos + 1, semi - pos - 1)
        : line.substr(pos + 1);
      try { return std::stoi(val_str); } catch (...) { return -1; }
    };

    if (line.rfind("AMIN:", 0) == 0) {
      int v = value_of("AMIN:");
      if (v >= 0) sensor_step_min_ = v;
    } else if (line.rfind("AMAX:", 0) == 0) {
      int v = value_of("AMAX:");
      if (v >= 0) sensor_step_max_ = v;
    } else if (line.rfind("ARES:", 0) == 0) {
      int v = value_of("ARES:");
      if (v > 0) angular_resolution_ = (2.0 * M_PI) / v;
    } else if (line.rfind("SCAN:", 0) == 0) {
      int v = value_of("SCAN:");
      if (v > 0) scan_time_ = 60.0 / v;
    }
  }

  apply_angle_limits();

  int num_steps = sensor_step_max_ - sensor_step_min_;
  RCLCPP_INFO(get_logger(),
    "  Sensor: steps %d–%d (%d points), scan time %.1f ms",
    sensor_step_min_, sensor_step_max_, num_steps, scan_time_ * 1000.0);
  RCLCPP_INFO(get_logger(),
    "  Requested scan: steps %d–%d, angles %.1f to %.1f deg",
    step_min_, step_max_,
    scan_angle_min_ * 180.0 / M_PI,
    scan_angle_max_ * 180.0 / M_PI);
}

void HokuyoScipDriver::apply_angle_limits()
{
  const double min_angle = std::min(angle_min_, angle_max_);
  const double max_angle = std::max(angle_min_, angle_max_);
  const double center_step =
    0.5 * (static_cast<double>(sensor_step_min_) + static_cast<double>(sensor_step_max_));

  int requested_min = static_cast<int>(std::lround(center_step + min_angle / angular_resolution_));
  int requested_max = static_cast<int>(std::lround(center_step + max_angle / angular_resolution_));

  requested_min = std::clamp(requested_min, sensor_step_min_, sensor_step_max_);
  requested_max = std::clamp(requested_max, sensor_step_min_, sensor_step_max_);

  if (requested_max <= requested_min) {
    RCLCPP_WARN(get_logger(),
      "Invalid angle limits %.3f..%.3f rad; using full sensor FOV",
      angle_min_, angle_max_);
    requested_min = sensor_step_min_;
    requested_max = sensor_step_max_;
  }

  step_min_ = requested_min;
  step_max_ = requested_max;
  scan_angle_min_ = (static_cast<double>(step_min_) - center_step) * angular_resolution_;
  scan_angle_max_ = (static_cast<double>(step_max_) - center_step) * angular_resolution_;
}

// ═════════════════════════════════════════════════════════════════════════════
//  Continuous Streaming (MD command)
// ═════════════════════════════════════════════════════════════════════════════

void HokuyoScipDriver::start_streaming()
{
  // MD command: MDsssseeeecc0ii\n
  char cmd[32];
  std::snprintf(cmd, sizeof(cmd), "MD%04d%04d%02d%01d00\n",
                step_min_, step_max_, cluster_, skip_);

  RCLCPP_INFO(get_logger(), "Starting continuous stream: %s", cmd);
  send_command(cmd);

  std::this_thread::sleep_for(std::chrono::milliseconds(100));
  const auto status = read_response();

  // Parse status (second line, first 2 chars)
  std::istringstream ss(status);
  std::string line0, line1;
  std::getline(ss, line0);
  if (std::getline(ss, line1) && line1.size() >= 2) {
    std::string code = line1.substr(0, 2);
    if (code == "00" || code == "99") {
      RCLCPP_INFO(get_logger(), "Continuous streaming started (40 Hz)");
    } else {
      RCLCPP_WARN(get_logger(), "MD status: %s", code.c_str());
    }
  }

  // Launch receive thread
  running_ = true;

  // Set recv timeout for the streaming loop (1 s)
  struct timeval tv{1, 0};
  ::setsockopt(sock_fd_, SOL_SOCKET, SO_RCVTIMEO, &tv, sizeof(tv));

  recv_thread_ = std::thread(&HokuyoScipDriver::receive_loop, this);
}

void HokuyoScipDriver::receive_loop()
{
  std::string buf;
  buf.reserve(16384);
  char chunk[8192];

  while (running_ && rclcpp::ok()) {
    ssize_t n = ::recv(sock_fd_, chunk, sizeof(chunk), 0);
    if (n < 0) {
      if (errno == EAGAIN || errno == EWOULDBLOCK) continue;  // timeout
      if (running_) {
        RCLCPP_ERROR(get_logger(), "recv error: %s", std::strerror(errno));
      }
      break;
    }
    if (n == 0) {
      RCLCPP_ERROR(get_logger(), "Connection lost");
      break;
    }

    buf.append(chunk, static_cast<size_t>(n));

    // Process complete messages (delimited by "\n\n")
    std::string::size_type pos;
    while ((pos = buf.find("\n\n")) != std::string::npos) {
      std::string message = buf.substr(0, pos);
      buf.erase(0, pos + 2);
      const auto rx_stamp = now();
      process_md_response(message, rx_stamp);
    }
  }
}

// ═════════════════════════════════════════════════════════════════════════════
//  SCIP 2.0 Decode
// ═════════════════════════════════════════════════════════════════════════════

std::vector<int> HokuyoScipDriver::scip_decode(const std::string & data)
{
  // Each line ends with a checksum character; strip it, then concatenate
  // everything and decode groups of 3 characters (6 bits each).
  std::string clean;
  clean.reserve(data.size());

  std::string::size_type start = 0;
  while (start < data.size()) {
    auto lf = data.find('\n', start);
    std::string::size_type end = (lf != std::string::npos) ? lf : data.size();
    if (end > start + 1) {  // at least 1 real char + checksum
      clean.append(data, start, end - start - 1);  // strip trailing checksum
    }
    start = (lf != std::string::npos) ? lf + 1 : data.size();
  }

  std::vector<int> values;
  values.reserve(clean.size() / 3);

  for (std::string::size_type i = 0; i + 3 <= clean.size(); i += 3) {
    int val = 0;
    for (int j = 0; j < 3; ++j) {
      val = (val << 6) + (static_cast<int>(clean[i + j]) - 0x30);
    }
    values.push_back(val);
  }
  return values;
}

// ═════════════════════════════════════════════════════════════════════════════
//  MD Response → LaserScan
// ═════════════════════════════════════════════════════════════════════════════

void HokuyoScipDriver::process_md_response(
  const std::string & message,
  const rclcpp::Time & rx_stamp)
{
  // Split into lines
  std::vector<std::string> lines;
  {
    std::istringstream ss(message);
    std::string l;
    while (std::getline(ss, l)) {
      // Remove trailing CR
      if (!l.empty() && l.back() == '\r') l.pop_back();
      lines.push_back(std::move(l));
    }
  }
  if (lines.size() < 4) return;

  // Line 0: echo of MD command
  if (lines[0].rfind("MD", 0) != 0) return;

  // Line 1: status (first 2 chars)
  if (lines[1].size() < 2 || lines[1].substr(0, 2) != "99") return;

  // Line 2: sensor timestamp (not used directly here).
  // We anchor scan.header.stamp at the driver receive time (rx_stamp)
  // so downstream latency monitors include in-driver handling.

  // Lines 3+: distance data
  std::string data_str;
  for (size_t i = 3; i < lines.size(); ++i) {
    if (!lines[i].empty()) {
      data_str += lines[i];
      data_str += '\n';
    }
  }
  if (data_str.empty()) return;

  const auto distances_mm = scip_decode(data_str);
  if (distances_mm.empty()) return;

  // Build LaserScan
  sensor_msgs::msg::LaserScan scan;
  scan.header.stamp    = rx_stamp;
  scan.header.frame_id = frame_id_;

  const int    num_points           = static_cast<int>(distances_mm.size());
  const double effective_resolution = angular_resolution_ * cluster_;

  scan.angle_min       = static_cast<float>(scan_angle_min_);
  scan.angle_max       = static_cast<float>(scan_angle_max_);
  scan.angle_increment = static_cast<float>(effective_resolution);
  scan.time_increment  = static_cast<float>(scan_time_ / num_points);
  scan.scan_time       = static_cast<float>(scan_time_);
  scan.range_min       = static_cast<float>(range_min_);
  scan.range_max       = static_cast<float>(range_max_);

  scan.ranges.resize(static_cast<size_t>(num_points));
  const float inf = std::numeric_limits<float>::infinity();
  const float rmin = static_cast<float>(range_min_);
  const float rmax = static_cast<float>(range_max_);

  for (int i = 0; i < num_points; ++i) {
    const int d = distances_mm[i];
    if (d <= 20) {          // invalid reading
      scan.ranges[i] = inf;
    } else {
      const float r = static_cast<float>(d) * 0.001f;  // mm → m
      scan.ranges[i] = (r < rmin || r > rmax) ? inf : r;
    }
  }

  scan_pub_->publish(scan);
}

}  // namespace f1tenth_lidar

// ═════════════════════════════════════════════════════════════════════════════
//  Main
// ═════════════════════════════════════════════════════════════════════════════
int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<f1tenth_lidar::HokuyoScipDriver>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
